pkg/providers/v2/instances_test.go

```diff
@@ -0,0 +1,431 @@
+						},
+					},
+				}
+			}
+
+			mockEC2.EXPECT().DescribeInstances(gomock.Any()).Return(expectedEc2Output, nil)
+
+			exists, err := fakeInstances.InstanceExists(context.TODO(), test.node)
+
+			if err != nil {
+				t.Errorf("InstanceExists failed with node %v: %v", nodeName, err)
+			}
+
+			if exists != test.expectedResult {
+				t.Errorf("unexpected result, InstanceExists should return %v", test.expectedResult)
+			}
+		})
+	}
+}
+
+func TestInstanceShutdown(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockEC2 := mocks.NewMockEC2(mockCtrl)
+
+	fakeInstances := &instances{
+		ec2: mockEC2,
+	}
+
+	nodeName := "ip-192-168-0-1.ec2.internal"
+
+	tests := []struct {
+		name            string
+		fakeEc2Instance ec2.Instance
+		node            *v1.Node
+		expectedResult  bool
+	}{
+		{
+			name:            "test InstanceShutdown with running instance",
+			fakeEc2Instance: makeInstance(0, "192.168.0.1", "1.2.3.6", "instance-same.ec2.internal", "instance-same.ec2.external", "running"),
+			node:            makeNode(nodeName),
+			expectedResult:  false,
+		},
+		{
+			name:            "test InstanceShutdown with running instance (node without providerID)",
+			fakeEc2Instance: makeInstance(0, "192.168.0.1", "1.2.3.6", "instance-same.ec2.internal", "instance-same.ec2.external", "running"),
+			node:            makeNodeWithoutProviderID(nodeName),
+			expectedResult:  false,
+		},
+		{
+			name:            "test InstanceShutdown with stopping instance",
+			fakeEc2Instance: makeInstance(0, "192.168.0.1", "1.2.3.6", "instance-same.ec2.internal", "instance-same.ec2.external", "stopping"),
+			node:            makeNode(nodeName),
+			expectedResult:  true,
+		},
+		{
+			name:            "test InstanceShutdown with stopped instance (node without providerID)",
+			fakeEc2Instance: makeInstance(0, "192.168.0.1", "1.2.3.6", "instance-same.ec2.internal", "instance-same.ec2.external", "stopped"),
+			node:            makeNodeWithoutProviderID(nodeName),
+			expectedResult:  true,
+		},
+		{
+			name:            "test InstanceShutdown with terminated instance",
+			fakeEc2Instance: makeInstance(0, "192.168.0.1", "1.2.3.6", "instance-same.ec2.internal", "instance-same.ec2.external", "terminated"),
+			node:            makeNode(nodeName),
+			expectedResult:  true,
+		},
+		{
+			name:            "test InstanceShutdown with terminated instance (node without provierID)",
+			fakeEc2Instance: makeInstance(0, "192.168.0.1", "1.2.3.6", "instance-same.ec2.internal", "instance-same.ec2.external", "terminated"),
+			node:            makeNodeWithoutProviderID(nodeName),
+			expectedResult:  true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			expectedEc2Output := &ec2.DescribeInstancesOutput{}
+			state := aws.StringValue(test.fakeEc2Instance.State.Name)
+			if state != ec2.InstanceStateNameTerminated {
+				expectedEc2Output = &ec2.DescribeInstancesOutput{
```

Comment:
Wondering if the entire `ec2.DescribeInstancesOutput` should be part of the test table instead of just `ec2.Instance`. This way we can differe the output based on the state

Reply:
yeah I've considered that way before, that should also work, but I feel this way can save a lot more space
I think using the other way is probably more readable? Will update :)